### DISCUSSION TITLE
confusion about shape inference

### DISCUSSION THREAD

**Discussion by Channingss** [onnx/onnx#2819] [Q&A]
1. Is Mul have shape inference? I don't find function:`TypeAndShapeInferenceFunction` of Mul, of course some other OPs too, which means some OPs don't have realizition of  shape inference?
https://github.com/onnx/onnx/blob/925b3657924c0c16cd20b54595f41e76159b03ab/onnx/defs/math/defs.cc#L179
2. I export a Bert model from this repo:https://github.com/huggingface/transformers/blob/master/notebooks/04-onnx-export.ipynb
I found this input of model is symbolic(not Int64), through input tensor is given by torch.onnx,export, and shape inference of some node is also failed, witch means onnx.shape_inference don't support symbolic shape, right?
![image](https://user-images.githubusercontent.com/12471701/84225300-62498300-ab11-11ea-93c7-a629f5e6d2b5.png)

**Comment by TMVector** [comment]
Hi @Channingss, on the line you point to you can see the `Mul` op is defined almost entirely in the `MathDocGenerator` function, which is where the type/shape inference code is:
https://github.com/onnx/onnx/blob/925b3657924c0c16cd20b54595f41e76159b03ab/onnx/defs/math/defs.cc#L32-L39

Shape inference does support named (non-concrete) dimensions, but you're right it doesn't do any symbolic manipulation (e.g. 2*x). However, ONNX Runtime has a [shape inference script](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/providers/nuphar/scripts/symbolic_shape_infer.py) for their Nuphar provider which _does_ do symbolic shape inference, so you can use that if you need.

**Comment by Channingss** [comment]
thanks for your reply,   I export a Bert model from this repo:https://github.com/huggingface/transformers/blob/master/notebooks/04-onnx-export.ipynb
why this input of onnx model is symbolic(not Int64), through input shape is fixed for torch.onnx.export
![image](https://user-images.githubusercontent.com/12471701/84340789-11e72980-abd4-11ea-8951-6688c5e642ec.png)

**Comment by TMVector** [comment]
If you dig into the code a little you can see they are providing a list of `dynamic_axes`, which tell PyTorch which axes of which inputs should be treated as named dimensions instead of fixed dimensions. You should still be able to run inference exactly the same though.
https://github.com/huggingface/transformers/blob/699541c4b34479451c91b3c6c204d904f62bed83/src/transformers/convert_graph_to_onnx.py#L130

You can find the documentation for that function [here](https://pytorch.org/docs/stable/onnx.html#torch.onnx.export).

Does this cause any issues for you @Channingss, or are you just curious? 🙂
